cpp/src/arrow/compute/kernels/scalar_cast_string.cc

```diff
@@ -305,19 +310,198 @@ BinaryToBinaryCastExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* ou
       ctx, input, out->array_data().get());
 }
 
+// View -> Span
+template <typename O, typename I>
+enable_if_t<is_binary_view_like_type<I>::value && is_base_binary_type<O>::value, Status>
+BinaryToBinaryCastExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+  using OutputBuilderType = typename TypeTraits<O>::BuilderType;
+  const CastOptions& options = checked_cast<const CastState&>(*ctx->state()).options;
+  const ArraySpan& input = batch[0].array;
+
+  if constexpr (!I::is_utf8 && O::is_utf8) {
+    if (!options.allow_invalid_utf8) {
+      InitializeUTF8();
+      ArraySpanVisitor<I> visitor;
+      Utf8Validator validator;
+      RETURN_NOT_OK(visitor.Visit(input, &validator));
+    }
+  }
+
+  // XXX: a more efficient implementation that zero-copies the validity bitmap
+  // is possible, but requires a more complex implementation for building the
+  // offsets and data buffers
+  OutputBuilderType builder(options.to_type.GetSharedPtr(), ctx->memory_pool());
+  RETURN_NOT_OK(builder.Resize(input.length));
```

Comment:
Is it possible to presize the data as well, in addition to the offsets?

Reply:
Not without visiting the whole array to sum the sizes. It would for example be legal to have all N views reference the same 1MB span of a data buffer

Reply:
"visiting the whole array to sum the sizes" sounds fine to me (I think we do it in other places as well), it should be beneficial especially if the input is large.

Reply:
Done. I implemented a helper that sums all the view sizes including nulls (avoid branching). Allocating a little bit more of memory (because it can potentially include null views that are not empty) shouldn't be a problem.